sql/tests: TestRandomSyntaxGeneration failed

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/19470569?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/19470569?buildTab=artifacts#/) on release-23.2.24-rc @ [f28e82e7081e80c8fb8442a82b713309f3ddcf1c](https://github.com/cockroachdb/cockroach/commits/f28e82e7081e80c8fb8442a82b713309f3ddcf1c):


```
    rsg_test.go:113: Last executed (4): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (5): ANALYZE TIES . SESSION . CHARACTERISTICS
    rsg_test.go:113: Last executed (6): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (7): COPY INDEX_BEFORE_PAREN ( ident , ident , TREAT , ident ) FROM error
    rsg_test.go:113: Last executed (8): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (9): IMPORT INTO FAMILY . NATURAL ( INTERVAL . error , OUT . error ) ident DATA ( 'string' , FULL ) WITH OPTIONS ( GROUPS = ident )
    rsg_test.go:113: Last executed (10): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (11): REFRESH MATERIALIZED VIEW NO_FULL_SCAN . CURRENT_USER . FAMILY WITH NO DATA
    rsg_test.go:113: Last executed (12): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (13): DECLARE DAY BINARY INSENSITIVE NO SCROLL CURSOR WITHOUT HOLD FOR ( TABLE error )
    rsg_test.go:113: Last executed (14): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (15): COPY ident . EXPIRATION . TIMETZ TO STDOUT WITH NULL PLACEHOLDER
    rsg_test.go:113: Last executed (16): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (17): PREPARE GEOMETRYCOLLECTIONZ ( POINT error ARRAY [ 449 ] , INTERVAL DAY TO HOUR ARRAY [ 126 ] [ error , INTERVAL [ ] [ error , POLYGON error [ 214 ] [ error ) AS OPT PLAN 'string'
    rsg_test.go:113: Last executed (18): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (19): DROP OWNED BY SESSION_USER , PARTIAL , LINESTRING RESTRICT
    rsg_test.go:113: Last executed (20): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (21): UNLISTEN TABLESPACE . ASC
    rsg_test.go:113: Last executed (22): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (23): COMMENT ON CONSTRAINT POLYGON ON INDEX_AFTER_ORDER_BY_BEFORE_AT IS 'string'
    rsg_test.go:113: Last executed (24): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (25): DECLARE SURVIVE BINARY ASENSITIVE CURSOR FOR ( ( TABLE error ) )
    rsg_test.go:113: Last executed (26): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (27): PREPARE ident ( COLLATION . RIGHT . DECIMAL [ 470 ] [ error , POINT error [ ] ) AS RESUME ALL ident JOBS
    rsg_test.go:113: Last executed (28): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (29): PREPARE TIME ( RESTRICT . ANY . ident ARRAY [ 330 ] [ error , CASCADE ARRAY [ -184 ] [ error , INTERVAL ARRAY ) AS OPT PLAN 'string'
    rsg_test.go:113: Last executed (30): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (31): MOVE -442 IN ident
    rsg_test.go:113: Last executed (32): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (33): COMMENT ON COLUMN INDEX_AFTER_ORDER_BY_BEFORE_AT . EXTRACT . BIT IS NULL
    rsg_test.go:113: Last executed (34): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (35): REFRESH MATERIALIZED VIEW CONCURRENTLY FAMILY . RIGHT . NO_ZIGZAG_JOIN
    rsg_test.go:113: Last executed (36): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (37): PREPARE ident ( BIT ( 169 ) ARRAY , @ 205 ARRAY [ -87 ] [ error , CROSS . ATOMIC . DATA ARRAY [ 25 ] [ error ) AS OPT PLAN 'string'
    rsg_test.go:113: Last executed (38): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (39): COPY NOTHING_AFTER_RETURNING ( BOOLEAN , ident , CONNECTIONS , TIMETZ , VIEW , FORMAT , GREATEST , ident ) FROM error
    rsg_test.go:113: Last executed (40): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (41): COPY FAMILY . DELIMITER . COLLATE ( NOCONTROLCHANGEFEED ) TO error
    rsg_test.go:113: Last executed (42): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (43): FETCH ABSOLUTE 158 IN IFERROR
    rsg_test.go:113: Last executed (44): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (45): COPY NOTHING_AFTER_RETURNING ( TIME , SCROLL ) TO STDOUT FORCE NULL error FORCE NOT NULL error
    rsg_test.go:113: Last executed (46): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (47): DECLARE VIEWACTIVITY SCROLL CURSOR FOR ( ( ( ( TABLE error ) ) ) )
    rsg_test.go:113: Last executed (48): CREATE DATABASE IF NOT EXISTS ident
    rsg_test.go:113: Last executed (49): COPY FAMILY ( ident , CONVERSION ) TO error
    rsg_test.go:116: Currently active statements: map[]
    rsg_test.go:353: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxGeneration2497083233
--- FAIL: TestRandomSyntaxGeneration (301.91s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #143630 sql/tests: TestRandomSyntaxGeneration failed [C-test-failure O-robot P-3 T-sql-foundations branch-release-24.3]
- #132558 sql/tests: TestRandomSyntaxGeneration failed [CREATE TENANT statement is not valid after parsing then formatting] [C-test-failure O-robot P-2 T-db-server branch-release-24.1.6-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-49504

--- a/pkg/sql/parser/sql.y
+++ b/pkg/sql/parser/sql.y
@@ -11307,6 +11307,7 @@ index_def:
         Predicate:        $11.expr(),
         Invisibility:     $12.indexInvisibility(),
       },
+      FormatAsIndex:    true,
     }
   }
 | INVERTED INDEX_BEFORE_PAREN '(' index_params ')' opt_partition_by_index opt_with_storage_parameter_list opt_where_clause opt_index_visible

--- a/pkg/sql/parser/testdata/create_table
+++ b/pkg/sql/parser/testdata/create_table
@@ -555,35 +555,35 @@ CREATE TABLE _ (_ INT8 UNIQUE) -- identifiers removed
 parse
 CREATE TABLE a (b INT, UNIQUE INDEX foo (b))
 ----
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b)) -- normalized!
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b)) -- fully parenthesized
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b)) -- literals removed
-CREATE TABLE _ (_ INT8, CONSTRAINT _ UNIQUE (_)) -- identifiers removed
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b)) -- normalized!
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b)) -- fully parenthesized
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b)) -- literals removed
+CREATE TABLE _ (_ INT8, UNIQUE INDEX _ (_)) -- identifiers removed
 
 parse
 CREATE TABLE a (b INT, UNIQUE INDEX foo (b) WHERE c > 3)
 ----
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b) WHERE c > 3) -- normalized!
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b) WHERE ((c) > (3))) -- fully parenthesized
-CREATE TABLE a (b INT8, CONSTRAINT foo UNIQUE (b) WHERE c > _) -- literals removed
-CREATE TABLE _ (_ INT8, CONSTRAINT _ UNIQUE (_) WHERE _ > 3) -- identifiers removed
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b) WHERE c > 3) -- normalized!
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b) WHERE ((c) > (3))) -- fully parenthesized
+CREATE TABLE a (b INT8, UNIQUE INDEX foo (b) WHERE c > _) -- literals removed
+CREATE TABLE _ (_ INT8, UNIQUE INDEX _ (_) WHERE _ > 3) -- identifiers removed
 
 parse
 CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST (c) (PARTITION d VALUES IN (1)))
 ----
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST (c) (PARTITION d VALUES IN (1))) -- normalized!
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST (c) (PARTITION d VALUES IN ((1)))) -- fully parenthesized
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST (c) (PARTITION d VALUES IN (_))) -- literals removed
-CREATE TABLE _ (UNIQUE (_) PARTITION BY LIST (_) (PARTITION _ VALUES IN (1))) -- identifiers removed
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST (c) (PARTITION d VALUES IN (1)))
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST (c) (PARTITION d VALUES IN ((1)))) -- fully parenthesized
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST (c) (PARTITION d VALUES IN (_))) -- literals removed
+CREATE TABLE _ (UNIQUE INDEX (_) PARTITION BY LIST (_) (PARTITION _ VALUES IN (1))) -- identifiers removed
 
 # Regression test for #95238
 parse
 CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN (1)))
 ----
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN (1))) -- normalized!
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN ((1)))) -- fully parenthesized
-CREATE TABLE a (UNIQUE (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN (_))) -- literals removed
-CREATE TABLE _ (UNIQUE (_) PARTITION BY LIST (_) (PARTITION _ VALUES IN (1))) -- identifiers removed
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN (1)))
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN ((1)))) -- fully parenthesized
+CREATE TABLE a (UNIQUE INDEX (b) PARTITION BY LIST ("c d") (PARTITION "e f" VALUES IN (_))) -- literals removed
+CREATE TABLE _ (UNIQUE INDEX (_) PARTITION BY LIST (_) (PARTITION _ VALUES IN (1))) -- identifiers removed
 
 parse
 CREATE TABLE a (b INT8 UNIQUE WITHOUT INDEX)
@@ -2639,3 +2639,35 @@ CREATE TABLE a (a VECTOR)
 CREATE TABLE a (a VECTOR) -- fully parenthesized
 CREATE TABLE a (a VECTOR) -- literals removed
 CREATE TABLE _ (_ VECTOR) -- identifiers removed
+
+parse
+CREATE TABLE t (a INT PRIMARY KEY, b INT, UNIQUE INDEX ( b NULLS FIRST ) USING HASH WITH ( 'foo' = 'bar' ))
+----
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH ('foo' = 'bar')) -- normalized!
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH ('foo' = ('bar'))) -- fully parenthesized
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH ('foo' = '_')) -- literals removed
+CREATE TABLE _ (_ INT8 PRIMARY KEY, _ INT8, UNIQUE INDEX (_ NULLS FIRST) USING HASH WITH ('foo' = 'bar')) -- identifiers removed
+
+parse
+CREATE TABLE t (a INT PRIMARY KEY, b INT, UNIQUE INDEX ( b NULLS FIRST ) USING HASH WITH BUCKET_COUNT = 10 WITH ( 'foo' = 'bar' ))
+----
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH BUCKET_COUNT = 10 WITH ('foo' = 'bar')) -- normalized!
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH BUCKET_COUNT = (10) WITH ('foo' = ('bar'))) -- fully parenthesized
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b NULLS FIRST) USING HASH WITH BUCKET_COUNT = _ WITH ('foo' = '_')) -- literals removed
+CREATE TABLE _ (_ INT8 PRIMARY KEY, _ INT8, UNIQUE INDEX (_ NULLS FIRST) USING HASH WITH BUCKET_COUNT = 10 WITH ('foo' = 'bar')) -- identifiers removed
+
+parse
+CREATE TABLE t (a INT PRIMARY KEY, b INT, UNIQUE INDEX (b) USING HASH)
+----
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b) USING HASH) -- normalized!
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b) USING HASH) -- fully parenthesized
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX (b) USING HASH) -- literals removed
+CREATE TABLE _ (_ INT8 PRIMARY KEY, _ INT8, UNIQUE INDEX (_) USING HASH) -- identifiers removed
+
+parse
+CREATE TABLE t (a INT PRIMARY KEY, b INT, UNIQUE INDEX idx (b) WITH ( 'foo' = 'bar' ))
+----
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX idx (b) WITH ('foo' = 'bar')) -- normalized!
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX idx (b) WITH ('foo' = ('bar'))) -- fully parenthesized
+CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, UNIQUE INDEX idx (b) WITH ('foo' = '_')) -- literals removed
+CREATE TABLE _ (_ INT8 PRIMARY KEY, _ INT8, UNIQUE INDEX _ (_) WITH ('foo' = 'bar')) -- identifiers removed

--- a/pkg/sql/sem/tree/create.go
+++ b/pkg/sql/sem/tree/create.go
@@ -1112,6 +1112,10 @@ type UniqueConstraintTableDef struct {
 	PrimaryKey   bool
 	WithoutIndex bool
 	IfNotExists  bool
+	// FormatAsIndex indicates if the constraint should be formatted as an index
+	// definition. This is needed since indexes support syntax for things like
+	// storage parameters and sharding, while constraints do not.
+	FormatAsIndex bool
 }
 
 // SetName implements the TableDef interface.
@@ -1126,7 +1130,7 @@ func (node *UniqueConstraintTableDef) SetIfNotExists() {
 
 // Format implements the NodeFormatter interface.
 func (node *UniqueConstraintTableDef) Format(ctx *FmtCtx) {
-	if node.Name != "" {
+	if node.Name != "" && !node.FormatAsIndex {
 		ctx.WriteString("CONSTRAINT ")
 		if node.IfNotExists {
 			ctx.WriteString("IF NOT EXISTS ")
@@ -1138,6 +1142,13 @@ func (node *UniqueConstraintTableDef) Format(ctx *FmtCtx) {
 		ctx.WriteString("PRIMARY KEY ")
 	} else {
 		ctx.WriteString("UNIQUE ")
+		if node.FormatAsIndex {
+			ctx.WriteString("INDEX ")
+			if node.Name != "" {
+				ctx.FormatNode(&node.Name)
+				ctx.WriteByte(' ')
+			}
+		}
 	}
 	if node.WithoutIndex {
 		ctx.WriteString("WITHOUT INDEX ")

--- a/pkg/sql/sem/tree/pretty.go
+++ b/pkg/sql/sem/tree/pretty.go
@@ -1810,12 +1810,19 @@ func (node *UniqueConstraintTableDef) doc(p *PrettyCfg) pretty.Doc {
 		if node.WithoutIndex {
 			title = pretty.ConcatSpace(title, pretty.Keyword("WITHOUT INDEX"))
 		}
+		if node.FormatAsIndex {
+			title = pretty.ConcatSpace(title, pretty.Keyword("INDEX"))
+		}
+	}
+	if node.Name != "" {
+		if node.FormatAsIndex {
+			title = pretty.ConcatSpace(title, p.Doc(&node.Name))
+		} else {
+			constraint := pretty.ConcatSpace(pretty.Keyword("CONSTRAINT"), p.Doc(&node.Name))
+			title = pretty.ConcatSpace(constraint, title)
+		}
 	}
 	title = pretty.ConcatSpace(title, p.bracket("(", p.Doc(&node.Columns), ")"))
-	if node.Name != "" {
-		clauses = append(clauses, title)
-		title = pretty.ConcatSpace(pretty.Keyword("CONSTRAINT"), p.Doc(&node.Name))
-	}
 	if node.Sharded != nil {
 		clauses = append(clauses, p.Doc(node.Sharded))
 	}

--- a/pkg/testutils/sqlutils/parse.go
+++ b/pkg/testutils/sqlutils/parse.go
@@ -70,8 +70,6 @@ func parseOne(t *testing.T, input string, p Parser) (tree.NodeFormatter, error) 
 func VerifyParseFormat(
 	t *testing.T, input, pos string, p Parser, reParseWithoutLiterals bool,
 ) string {
-	t.Helper()
-
 	// Check parse.
 	stmts, err := parse(t, input, p)
 	if err != nil {

--- a/pkg/testutils/sqlutils/pretty.go
+++ b/pkg/testutils/sqlutils/pretty.go
@@ -15,8 +15,6 @@ import (
 // VerifyStatementPrettyRoundtrip verifies that the SQL statements in s
 // correctly round trip through the pretty printer.
 func VerifyStatementPrettyRoundtrip(t *testing.T, sql string) {
-	t.Helper()
-
 	stmts, err := parser.Parse(sql)
 	if err != nil {
 		t.Fatalf("%s: %s", err, sql)
@@ -41,7 +39,6 @@ func VerifyStatementPrettyRoundtrip(t *testing.T, sql string) {
 func verifyStatementPrettyRoundTrip(
 	t *testing.T, sql string, origStmt tree.NodeFormatter, p Parser,
 ) {
-	t.Helper()
 	// Dataflow of the statement through these checks:
 	//
 	//             sql (from test file)

